dcos_test_utils/dcos_api.py

```diff
@@ -377,19 +377,27 @@ def _wait_for_srouter_slaves_endpoints(self):
         slaves_ids = sorted(x['id'] for x in data['slaves'] if x['hostname'] in self.all_slaves)
 
         for slave_id in slaves_ids:
-            # AdminRouter's slave endpoint internally uses cached Mesos
-            # state data. That is, slave IDs of just recently joined
-            # slaves can be unknown here. For those, this endpoint
-            # returns a 404. Retry in this case, until this endpoint
-            # is confirmed to work for all known agents.
+            in_progress_status_codes = (
+                # AdminRouter's slave endpoint internally uses cached Mesos
+                # state data. That is, slave IDs of just recently joined
+                # slaves can be unknown here. For those, this endpoint
+                # returns a 404. Retry in this case, until this endpoint
+                # is confirmed to work for all known agents.
+                404,
+                # We have seen 502 responses that have been followed by 200
+                # responses during an upgrade.
+                # This is likely for the same reason as the 404 responses
+                # described above.
+                502,
```

Comment:
A 502 is not of the same cause, it's the reverse proxy's (nginx in ourcase) method to tell the user agent that it right now cannot reach the backend (Mesos agent HTTP server) for transport layer reasons.

Reply:
5xx errors in general mean "transient error" and we want to retry. So: code behavior is good, comment is not ideal.

Reply:
Thanks both - I have updated the comment now.

Reply:
The 502 means that nginx _knows_ where to connect to (agent ID known), but that connection does not work (transport layer problem far down in the stack).

The 404 means that nginx does not know where to connect to (agend ID not known, application level problem).